### ISSUE TITLE
[Bug Report] IOIDataset double-prepends BOS and ignores prepend_bos=False

### ISSUE THREAD

**Issue by emerardd** [TransformerLensOrg/TransformerLens#1772]
**Describe the bug**

`IOIDataset.__getitem__()` calls `tokenizer.encode()` with its default special-token behavior and then manually prepends `bos_token_id` when `prepend_bos=True`.

TransformerLens tokenizers configured with `add_bos_token=True` therefore produce two leading BOS tokens on the default path. With `prepend_bos=False`, `encode()` still adds one BOS, so the flag does not disable BOS insertion.

The same unqualified `encode()` calls are used for the `IO` and `S` labels. Their shared automatic BOS shifts the first differing label-token index, and together with the prompt's second BOS causes the default `ioi_eval()` path to read logits one position later than intended.

**Code example**

This network-free reproduction uses a tokenizer stub with the same `add_special_tokens` contract:

```python
from unittest.mock import MagicMock

from transformer_lens.evals import IOIDataset

tokenizer = MagicMock()
tokenizer.bos_token_id = 0

def encode(text, add_special_tokens=True):
    payload = [11, 12, 21] if len(text) > 7 else [21 if "Alice" in text else 22]
    return ([0] if add_special_tokens else []) + payload

tokenizer.encode.side_effect = encode

for prepend_bos in (True, False):
    dataset = IOIDataset(
        tokenizer,
        templates=["[A] gave [B] to [A]"],
        names=["Alice", "Bob"],
        nouns={},
        num_samples=1,
        prepend_bos=prepend_bos,
        seed=0,
    )
    item = dataset[0]
    print(prepend_bos, item["prompt"].tolist(), item["IO"].tolist(), item["S"].tolist())
```

On current `dev`:

```text
True  [0, 0, 11, 12, 21] [0, 22] [0, 21]
False [0, 11, 12, 21]    [0, 22] [0, 21]
```

Expected:

- `prepend_bos=True`: exactly one leading BOS in `prompt`.
- `prepend_bos=False`: no leading BOS in `prompt`.
- `IO` and `S`: no automatically inserted special tokens, so their first differing content token is used directly.

A focused fix would encode all three fields with `add_special_tokens=False`, then let `prepend_bos` be the only code path that adds a prompt BOS.

**System Info**

- Installed from source with `uv`
- Base: `dev` at `e2581b5f3f3f62ae1fa8c770cb95adb187843fa4`
- OS: Windows NT 10.0.26200.0
- Python: 3.12.10

**Additional context**

The existing unit tokenizer stub always returns a fixed list and does not model automatic special-token insertion, so it cannot detect this contract violation. Regression coverage should exercise both values of `prepend_bos`, assert that label encodings omit special tokens, and verify the logit position used by `ioi_eval()`.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/TransformerLensOrg/TransformerLens/issues) in the repo

**Comment by emerardd** [comment]
I'll work on this.
